File: apps/web/src/lib/murmur/ssrf.ts
/**
 * SSRF allowlist + DNS-rebinding-aware fetcher for agent-supplied URLs.
 *
 * Used by every Murmur subcommand route in `apps/web/app/api/**` that
 * accepts a URL field (`board_url`, `sample_url`, ...). Two phases:
 *
 *   1. Host-pattern allowlist (synchronous, no DNS).
 *      Hard-coded list. Updates require a code change; no env-var
 *      override.
 *
 *   2. DNS resolution + IP filter (rebinding-aware).
 *      Resolve once; reject if the address is private, loopback,
 *      link-local, an IANA-reserved cloud-metadata IP, or an IPv6 ULA.
 *      `safeFetch` opens the connection to the *captured* IP so a second
 *      DNS lookup cannot redirect us to a private address.
 *
 * Allowlist scope: the initial set covers the demo's most common board
 * providers. Expand patterns as J5/#2759's routes encounter legitimate
 * rejections in real demo runs. The full reference list of board hosts
 * is at `apps/crawler/data/boards.csv`.
 *
 * Caller boundary: only files under `apps/web/app/api/**` (route
 * boundaries) and this module's own test file may import from here. The
 * `scripts/grep-validateurl-boundary.sh` gate enforces this.
 *
 * @see colophon-group/jobseek#2758
 * @see Murmur DESIGN.md §3.6 (Publisher SSRF defense), §4.2 (Probe SSRF defense)
 */

import { lookup as dnsLookup } from "node:dns/promises";
import type { LookupAddress } from "node:dns";

/**
 * Host-pattern allowlist. Each entry is either:
 *   - A literal host (e.g. `boards.greenhouse.io`) — exact match only.
 *   - A leading-wildcard pattern (e.g. `*.greenhouse.io`) — matches any
 *     subdomain at any depth, but NOT the bare apex (`greenhouse.io`).
 *
 * To add a host: edit this array. Code review is the gate; no runtime
 * override exists. Reference list of all known board hosts in use today:
 * `apps/crawler/data/boards.csv` (the `board_url` column).
 *
 * Expand as J5 routes encounter legitimate rejections in real demo runs.
 */
export const HOST_ALLOWLIST: readonly string[] = [
  // Greenhouse (most popular ATS in boards.csv)
  "*.greenhouse.io",
  "boards.greenhouse.io",
  "job-boards.greenhouse.io",
  // Lever
  "*.lever.co",
  "jobs.lever.co",
  // Workday
  "*.myworkdayjobs.com",
];

export type ValidateUrlOk = {
  ok: true;
  url: URL;
  resolvedIp: string;
  family: 4 | 6;
};

export type ValidateUrlErrCode =
  | "url_invalid"
  | "url_not_allowed"
  | "url_dns_failed"
  | "url_resolves_to_private";

export type ValidateUrlErr = {
  ok: false;
  error: ValidateUrlErrCode;
};

export type ValidateUrlResult = ValidateUrlOk | ValidateUrlErr;

/**
 * Match a hostname against a single allowlist pattern.
 *
 * Patterns:
 *   - `host.example.com` — exact match (case-insensitive).
 *   - `*.example.com` — any strict subdomain (one or more labels).
 *
 * No support for mid-string wildcards or path patterns; we deliberately
 * keep the matcher tiny so its behaviour is auditable in one glance.
 *
 * @param host    Hostname (no port, no path).
 * @param pattern Allowlist pattern.
 * @returns true iff `host` is allowed by `pattern`.
 */
export function matchHostPattern(host: string, pattern: string): boolean {
  const h = host.toLowerCase();
  const p = pattern.toLowerCase();

  if (!p.startsWith("*.")) {
    return h === p;
  }

  // `*.example.com` => suffix `.example.com` and at least one label
  // before it. The bare apex `example.com` does NOT match.
  const suffix = p.slice(1); // ".example.com"
  if (!h.endsWith(suffix)) return false;
  const prefix = h.slice(0, h.length - suffix.length);
  // prefix must be one-or-more non-empty labels (i.e. not empty and not
  // contain a leading dot).
  if (prefix.length === 0) return false;
  if (prefix.startsWith(".")) return false;
  return true;
}

/**
 * Internal helper: does `host` match ANY allowlist pattern?
 */
function isHostAllowed(host: string): boolean {
  for (const pattern of HOST_ALLOWLIST) {
    if (matchHostPattern(host, pattern)) return true;
  }
  return false;
}

/**
 * Internal: parse a dotted-quad IPv4 string into 4 octets, or null.
 */
function parseIPv4(ip: string): [number, number, number, number] | null {
  const parts = ip.split(".");
  if (parts.length !== 4) return null;
  const out: number[] = [];
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) return null;
    const n = Number(part);
    if (n < 0 || n > 255) return null;
    out.push(n);
  }
  return out as [number, number, number, number];
}

/**
 * Classify an IPv4 address as private/reserved.
 */
function isPrivateIPv4(ip: string): boolean {
  const parts = parseIPv4(ip);
  if (!parts) return false;
  const [a, b] = parts;

  // 0.0.0.0/8 unspecified / current network
  if (a === 0) return true;
  // 10.0.0.0/8 RFC1918
  if (a === 10) return true;
  // 100.64.0.0/10 CGNAT
  if (a === 100 && b >= 64 && b <= 127) return true;
  // 127.0.0.0/8 loopback
  if (a === 127) return true;
  // 169.254.0.0/16 link-local (includes metadata 169.254.169.254)
  if (a === 169 && b === 254) return true;
  // 172.16.0.0/12 RFC1918
  if (a === 172 && b >= 16 && b <= 31) return true;
  // 192.0.0.0/24 IETF protocol assignments
  if (a === 192 && b === 0 && parts[2] === 0) return true;
  // 192.0.2.0/24 TEST-NET-1
  if (a === 192 && b === 0 && parts[2] === 2) return true;
  // 192.168.0.0/16 RFC1918
  if (a === 192 && b === 168) return true;
  // 198.18.0.0/15 benchmarking
  if (a === 198 && (b === 18 || b === 19)) return true;
  // 198.51.100.0/24 TEST-NET-2
  if (a === 198 && b === 51 && parts[2] === 100) return true;
  // 203.0.113.0/24 TEST-NET-3
  if (a === 203 && b === 0 && parts[2] === 113) return true;
  // 224.0.0.0/4 multicast
  if (a >= 224 && a <= 239) return true;
  // 240.0.0.0/4 reserved (and 255.255.255.255 broadcast)
  if (a >= 240) return true;

  return false;
}

/**
 * Normalise an IPv6 string (lowercased, expanded `::`) into eight 16-bit
 * groups. Returns null on parse failure.
 */
function parseIPv6(ip: string): number[] | null {
  const lower = ip.toLowerCase();
  // IPv4-mapped form: ::ffff:1.2.3.4 — caller handles separately.
  if (lower.includes(".")) return null;

  const parts = lower.split("::");
  if (parts.length > 2) return null;

  const head = parts[0] === "" ? [] : parts[0].split(":");
  const tail = parts.length === 2 ? (parts[1] === "" ? [] : parts[1].split(":")) : [];

  if (parts.length === 1 && head.length !== 8) return null;
  if (parts.length === 2 && head.length + tail.length > 7) return null;

  const fillCount = 8 - head.length - tail.length;
  const groups: number[] = [];
  for (const g of head) {
    if (!/^[0-9a-f]{1,4}$/.test(g)) return null;
    groups.push(parseInt(g, 16));
  }
  for (let i = 0; i < fillCount; i++) groups.push(0);
  for (const g of tail) {
    if (!/^[0-9a-f]{1,4}$/.test(g)) return null;
    groups.push(parseInt(g, 16));
  }
  if (groups.length !== 8) return null;
  return groups;
}

/**
 * Classify an IPv6 address as private/reserved.
 */
function isPrivateIPv6(ip: string): boolean {
  const lower = ip.toLowerCase();

  // Unspecified `::`
  if (lower === "::" || lower === "0:0:0:0:0:0:0:0") return true;
  // Loopback `::1`
  if (lower === "::1" || lower === "0:0:0:0:0:0:0:1") return true;

  // IPv4-mapped IPv6: `::ffff:a.b.c.d` — re-classify the embedded v4.
  const v4MappedMatch = lower.match(/^::ffff:(\d+\.\d+\.\d+\.\d+)$/);
  if (v4MappedMatch) {
    return isPrivateIPv4(v4MappedMatch[1]);
  }

  const groups = parseIPv6(lower);
  if (!groups) return false;

  const first = groups[0];

  // fe80::/10 link-local (first 10 bits = 1111111010)
  if ((first & 0xffc0) === 0xfe80) return true;
  // fc00::/7 ULA (first 7 bits = 1111110)
  if ((first & 0xfe00) === 0xfc00) return true;
  // ff00::/8 multicast
  if ((first & 0xff00) === 0xff00) return true;
  // ::/8 reserved (covers IPv4-compat and other legacy)
  if ((first & 0xff00) === 0x0000 && groups.slice(1, 7).every((g) => g === 0)) {
    // ::1 already handled; ::ffff: handled; bare :: handled. Any other
    // ::xxxx is suspicious — treat as private to fail closed.
    return true;
  }

  return false;
}

/**
 * Classify an IP address as public or private/reserved.
 *
 * Treats as PRIVATE (returns true):
 *   - Loopback           (`127.0.0.0/8`, `::1`)
 *   - RFC1918            (`10/8`, `172.16/12`, `192.168/16`)
 *   - CGNAT              (`100.64/10`)
 *   - Link-local         (`169.254.0.0/16`, `fe80::/10`)
 *   - Metadata services  (`169.254.169.254`)
 *   - IPv6 ULA           (`fc00::/7`)
 *   - IPv4-mapped IPv6   (`::ffff:<v4>` — re-classify the embedded v4)
 *   - Unspecified        (`0.0.0.0`, `::`)
 *   - Multicast / broadcast / IETF-reserved ranges
 *
 * Anything else is public.
 */
export function isPrivateIp(ip: string): boolean {
  if (parseIPv4(ip)) return isPrivateIPv4(ip);
  return isPrivateIPv6(ip);
}

/**
 * Validate a URL against the SSRF allowlist + IP filter.
 *
 * Performs phase 1 (allowlist) synchronously and phase 2 (DNS) only if
 * phase 1 passes. Returns the resolved IP so the caller can route via a
 * rebinding-aware fetcher.
 *
 * Never throws — all error states are encoded in the return shape.
 *
 * Error codes:
 *   - `url_invalid`              — input was not a parseable absolute URL with `http(s):` scheme.
 *   - `url_not_allowed`          — host did not match any allowlist pattern. NO DNS LOOKUP IS PERFORMED.
 *   - `url_dns_failed`           — DNS resolution failed (NXDOMAIN, timeout, no A/AAAA record).
 *   - `url_resolves_to_private`  — resolved address is private, loopback, link-local, metadata, or ULA.
 */
export async function validateUrl(input: string): Promise<ValidateUrlResult> {
  // Phase 0: parse + scheme check.
  let url: URL;
  try {
    url = new URL(input);
  } catch {
    return { ok: false, error: "url_invalid" };
  }
  if (url.protocol !== "https:" && url.protocol !== "http:") {
    return { ok: false, error: "url_invalid" };
  }
  if (!url.hostname) {
    return { ok: false, error: "url_invalid" };
  }

  const host = url.hostname.toLowerCase();

  // Phase 1: allowlist match. Performs NO DNS lookup on miss.
  if (!isHostAllowed(host)) {
    return { ok: false, error: "url_not_allowed" };
  }

  // Phase 2: DNS resolution + IP filter.
  let resolved: LookupAddress;
  try {
    // `verbatim: false` asks Node to honour the system's address-family
    // preference; it does NOT relax filtering. We capture both the
    // address and the family so safeFetch can pin the connection.
    resolved = await dnsLookup(host, { verbatim: false });
  } catch {
    return { ok: false, error: "url_dns_failed" };
  }

  const family = resolved.family === 6 ? 6 : 4;
  if (isPrivateIp(resolved.address)) {
    return { ok: false, error: "url_resolves_to_private" };
  }

  return { ok: true, url, resolvedIp: resolved.address, family };
}

/**
 * Error thrown by `safeFetch` when validation fails.
 */
export class SafeFetchError extends Error {
  public readonly code: ValidateUrlErrCode;
  constructor(code: ValidateUrlErrCode, message?: string) {
    super(message ?? code);
    this.name = "SafeFetchError";
    this.code = code;
  }
}

/**
 * Fetch a URL after validation, with DNS-rebinding protection.
 *
 * Resolves the hostname *once* via `validateUrl` and captures the IP.
 * The outbound request is then issued against a URL whose hostname has
 * been rewritten to that captured IP, with the original Host header
 * preserved on the request so virtual-hosted servers route correctly.
 *
 * Why this works as rebinding mitigation: by substituting the IP into
 * the URL we prevent any further DNS resolution. The connect call has
 * no opportunity to consult the resolver again, so an attacker who
 * controls the hostname's DNS cannot redirect us to a private IP after
 * the validation check has passed.
 *
 * Trade-off: this approach breaks TLS hostname verification (the
 * server's certificate is for the original hostname, not the IP). For
 * the M0 demo all probe targets are public ATS providers; if/when this
 * is reused for richer transport, swap to a `node:https` Agent with a
 * custom `lookup` AND `servername` so SNI continues to use the original
 * hostname while the connection still routes to the captured IP.
 *
 * Returns whatever `fetch` produces. If validation fails, throws
 * `SafeFetchError` with `.code` matching `ValidateUrlErrCode`.
 *
 * Callers at route boundaries should still invoke `validateUrl` first
 * to surface a structured error to the agent; `safeFetch` is the
 * defence-in-depth wrapper for the actual outbound call.
 */
export async function safeFetch(
  input: string,
  init?: RequestInit,
): Promise<Response> {
  const v = await validateUrl(input);
  if (!v.ok) {
    throw new SafeFetchError(v.error);
  }

  // Defence in depth: re-check the captured IP. If anything has stuffed
  // a private IP into our captured value, fail closed before opening a
  // socket.
  if (isPrivateIp(v.resolvedIp)) {
    throw new SafeFetchError("url_resolves_to_private");
  }

  // Pin the connection to the captured IP by substituting it into the
  // URL hostname. Bracket IPv6 literals.
  const ipUrl = new URL(v.url.toString());
  ipUrl.hostname = v.family === 6 ? `[${v.resolvedIp}]` : v.resolvedIp;

  // Preserve the original Host header so virtual-hosted servers still
  // route the request to the right vhost.
  const headers = new Headers(init?.headers);
  if (!headers.has("host")) {
    headers.set("host", v.url.host);
  }

  return fetch(ipUrl, { ...init, headers });
}

File: apps/crawler/src/shared/ssrf.py
"""SSRF guard for board fetches (closes #3214).

The crawler fetches arbitrary URLs sourced from ``data/boards.csv`` and
the agent-driven workspace flow. Without a guard, a board URL such as
``http://10.0.0.5/admin`` (or any of the Hetzner private-network IPs)
would be followed straight to internal services. Worse: an attacker who
controls the DNS for a permitted hostname can swing the A record to a
private IP between the validation step and the actual connection — the
classic DNS-rebinding twist — and a redirect to a private IP slips past
any host-string check.

The guard is implemented at the **httpx transport layer**:

  1. :func:`is_private_ip` — pure classifier matching the CIDR set
     called out by the issue (10/8, 172.16/12, 192.168/16, 127/8,
     169.254/16, ::1/128, fc00::/7) plus the IETF-reserved ranges that
     are equally dangerous (unspecified, multicast, CGNAT, link-local
     v6, IPv4-mapped v6, metadata).
  2. :func:`resolve_host_or_raise` — runs ``socket.getaddrinfo`` and rejects
     when *any* returned address is private. The sole exception is an
     unscoped IPv6 link-local DNS answer alongside a public address: without
     an interface scope it is not a routable destination. The async transport
     retries only temporary ``EAI_AGAIN`` resolver failures with a bounded
     backoff; every successful answer is still validated fail-closed.
  3. :func:`is_internal_host_allowed` — opt-in exemption list, sourced
     from ``INTERNAL_HOSTS_ALLOW`` env (comma-separated host[:port]
     entries) plus the proxy / Typesense / Postgres / Redis hosts the
     deployment legitimately routes onto the private network.
  4. :class:`SSRFGuardedTransport` — an ``httpx.AsyncBaseTransport``
     wrapper that runs :func:`validate_request_url` before every
     ``handle_async_request`` call. httpx's redirect engine re-enters
     ``transport.handle_async_request`` for every hop, so this single
     hook catches both direct connections *and* redirect targets.

Threat model:

  - Untrusted board URL points to a private IP directly        → blocked
  - Untrusted host resolves to a private IP (rebinding)        → blocked
  - Permitted host that 30x-redirects to ``http://127.0.0.1/`` → blocked
  - Webshare proxy URL on the private network → NOT blocked
    (the proxy URL is supplied via httpx's ``proxy=`` kwarg, not the
    request URL, so it never reaches :meth:`handle_async_request` as
    the URL to validate)
  - Typesense / Postgres / Redis traffic                       → NOT blocked
    (these go through dedicated clients — ``typesense.Client``,
    ``asyncpg``, ``redis-py`` — never the shared httpx clients)

The proxy-routed transport (``apps/crawler/AGENTS.md`` "Proxy-routed
transport") legitimately uses HTTP proxies on the Hetzner private
network. We don't validate the proxy URL itself, only the request URL —
so private-IP proxies keep working. The internal-host allowlist is a
last-resort knob for, e.g., a local dev Typesense board URL during
testing; it is not required for the live deployment.

@see colophon-group/jobseek#3214
@see apps/murmur-shim/src/lib/murmur/ssrf.ts (TS sibling — same model)
"""

from __future__ import annotations

import asyncio
import ipaddress
import os
import socket
from typing import TYPE_CHECKING, Any
from urllib.parse import urlparse

import httpx
import structlog

if TYPE_CHECKING:
    from collections.abc import Iterable

log = structlog.get_logger()

DNS_RETRY_BACKOFF_SECONDS = (0.05, 0.1)


class SSRFError(Exception):
    """Raised when a URL fails the SSRF guard.

    Surfaces as an ``httpx.RequestError`` subclass-style exception
    propagated through ``handle_async_request``. Monitors / scrapers
    that ``except Exception`` already route this through their failure
    path (``_RECORD_FAILURE``); callers that need the structured
    code can ``isinstance(exc, SSRFError)``.
    """

    def __init__(self, url: str, reason: str, *, host: str | None = None) -> None:
        self.url = url
        self.host = host
        self.reason = reason
        super().__init__(f"SSRF guard refused {url!r}: {reason}")


def _ip_in_blocklist(ip: ipaddress.IPv4Address | ipaddress.IPv6Address) -> str | None:
    """Return a reason string when ``ip`` is in a blocked range, else ``None``.

    Mirrors the issue's required block list plus a handful of IANA-reserved
    ranges that ``ipaddress`` already classifies for us. Using the stdlib
    properties avoids re-encoding CIDR arithmetic in Python.
    """
    if ip.is_loopback:
        # 127.0.0.0/8 and ::1/128
        return "loopback"
    if ip.is_link_local:
        # 169.254.0.0/16 and fe80::/10. Covers cloud metadata
        # (169.254.169.254) and IPv6 link-local — never legitimate
        # destinations for a public-Internet fetch.
        return "link_local"
    if ip.is_private:
        # 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16 for v4; fc00::/7 for v6.
        # ``ipaddress.is_private`` also flags the unspecified address and a
        # few IETF-reserved ranges, which we want to block anyway.
        return "private"
    if ip.is_multicast:
        return "multicast"
    if ip.is_reserved:
        # 240.0.0.0/4, etc. — never a real Internet target.
        return "reserved"
    if ip.is_unspecified:
        # 0.0.0.0, :: — kernels route these to loopback / "any" on bind.
        return "unspecified"
    # IPv4-mapped IPv6: re-classify the embedded v4 so an attacker
    # can't smuggle 10.0.0.1 in as ``::ffff:10.0.0.1``.
    if isinstance(ip, ipaddress.IPv6Address) and ip.ipv4_mapped is not None:
        return _ip_in_blocklist(ip.ipv4_mapped)
    return None


def is_private_ip(addr: str) -> bool:
    """Whether ``addr`` is in any blocked range (private/loopback/etc).

    Public surface for tests and operator scripts. Accepts any v4 or
    v6 literal that ``ipaddress.ip_address`` understands. Unparseable
    input returns ``True`` (fail-closed) — an SSRF guard that lets a
    malformed address through is worse than one that rejects it.
    """
    try:
        ip = ipaddress.ip_address(addr)
    except ValueError:
        return True
    return _ip_in_blocklist(ip) is not None


def _normalize_host_entry(entry: str) -> str:
    """Normalise an allowlist entry — strip whitespace, lowercase, drop port."""
    entry = entry.strip().lower()
    if not entry:
        return ""
    # Support ``host`` and ``host:port`` — only the host portion matters
    # for matching, since the guard is keyed on what was resolved.
    if entry.startswith("["):
        # IPv6 literal in brackets: ``[fe80::1]:8108``
        end = entry.find("]")
        return entry[1:end] if end > 0 else entry.strip("[]")
    if ":" in entry and not entry.replace(":", "").replace(".", "").isalnum():
        # Probably a bare IPv6 literal like ``fe80::1``. Keep as-is.
        return entry
    return entry.split(":", 1)[0]


def _extract_host(url: str) -> str:
    """Pull the host out of ``url``. Empty string if unparseable."""
    if not url:
        return ""
    try:
        parsed = urlparse(url)
    except ValueError:
        return ""
    host = parsed.hostname or ""
    return host.lower()


def _gather_internal_hosts() -> frozenset[str]:
    """Hosts the deployment legitimately speaks to on the private network.

    Comes from three sources (union):

      1. ``INTERNAL_HOSTS_ALLOW`` env — comma-separated ``host[:port]``.
         Operator-facing override; required for a board URL that
         genuinely points at an internal service.
      2. The Webshare proxy URLs (``WEBSHARE_PROXY_URLS`` /
         ``WEBSHARE_PROXY_URL``).
         Defence-in-depth only: the proxy URL is supplied via httpx's
         ``proxy=`` kwarg and isn't the request URL, so it can't
         actually trip the guard — but if a future caller ever passes
         it as a request URL, it should still be allowed.
      3. ``TYPESENSE_HOST`` / ``LOCAL_DATABASE_URL`` / ``REDIS_URL``.
         These services do not go through ``src.shared.http`` today
         (they use ``typesense.Client``, ``asyncpg``, ``redis-py``),
         but seeding them here means an inadvertent future migration
         to an httpx-based client won't suddenly start blocking
         intra-cluster traffic.

    Read lazily so tests can monkeypatch ``settings`` / env without
    a module-import dance.
    """
    out: set[str] = set()

    # 1. Operator override — settings first, env as fallback so test
    # monkeypatching of ``settings.internal_hosts_allow`` works without
    # also mutating ``os.environ``. Both are read so a deploy-time
    # environment variable still wins when pydantic-settings hasn't
    # been re-instantiated (e.g. inside a long-running container after
    # an env change).
    raw_entries: list[str] = []
    try:
        from src.config import settings as _settings
    except Exception:
        _settings = None
    if _settings is not None:
        configured = getattr(_settings, "internal_hosts_allow", "") or ""
        if configured:
            raw_entries.append(configured)
    env_value = os.environ.get("INTERNAL_HOSTS_ALLOW", "")
    if env_value:
        raw_entries.append(env_value)

    for raw in raw_entries:
        for entry in raw.split(","):
            host = _normalize_host_entry(entry)
            if host:
                out.add(host)

    # 2 + 3. Deployment-derived. ``src.config`` may not be importable in
    # test contexts that haven't seeded ``DATABASE_URL``; swallow the
    # error and continue with whatever we have.
    if _settings is None:
        return frozenset(out)

    deployment_urls = [
        *getattr(_settings, "webshare_proxy_urls", []),
        getattr(_settings, "webshare_proxy_url", ""),
        getattr(_settings, "local_database_url", ""),
        getattr(_settings, "database_url", ""),
        getattr(_settings, "redis_url", ""),
    ]
    for url in deployment_urls:
        host = _extract_host(url)
        if host:
            out.add(host)

    ts_host = getattr(_settings, "typesense_host", "")
    if ts_host:
        host = _normalize_host_entry(ts_host)
        if host:
            out.add(host)

    return frozenset(out)


def is_internal_host_allowed(host: str) -> bool:
    """Whether ``host`` is in the deployment's internal-services allowlist."""
    h = host.strip().lower()
    if not h:
        return False
    return h in _gather_internal_hosts()


def _public_address_from_infos(
    url: str,
    host: str,
    infos: Iterable[tuple[Any, ...]],
) -> str:
    """Validate one resolver answer and choose its first public address."""
    chosen: str | None = None
    unscoped_link_local: str | None = None
    for family, _socktype, _proto, _canonname, sockaddr in infos:
        has_textual_scope = False
        if family == socket.AF_INET:
            addr = sockaddr[0]
        elif family == socket.AF_INET6:
            addr = sockaddr[0]
            # Strip the zone-id suffix some IPv6 link-local entries
            # carry (``fe80::1%en0``) before classifying.
            has_textual_scope = "%" in addr
            addr = addr.split("%", 1)[0]
        else:
            continue
        if is_private_ip(addr):
            ip = ipaddress.ip_address(addr)
            scope_id = sockaddr[3] if family == socket.AF_INET6 and len(sockaddr) > 3 else 0
            if (
                isinstance(ip, ipaddress.IPv6Address)
                and ip.is_link_local
                and not scope_id
                and not has_textual_scope
            ):
                # A DNS AAAA record cannot supply the interface scope required
                # to route an IPv6 link-local destination. Linux rejects a
                # connect to this sockaddr with EINVAL. Some public sites
                # nevertheless publish such broken AAAA records alongside a
                # valid A record, so defer this one unreachable answer and use
                # the public address when available. Scoped link-local answers
                # remain blocked because they identify a reachable interface.
                unscoped_link_local = addr
                continue
            raise SSRFError(
                url,
                f"resolves_to_private_ip:{addr}",
                host=host,
            )
        if chosen is None:
            chosen = addr

    if chosen is None and unscoped_link_local is not None:
        raise SSRFError(
            url,
            f"resolves_to_private_ip:{unscoped_link_local}",
            host=host,
        )
    if chosen is None:
        raise SSRFError(url, "no_address_record", host=host)
    return chosen


def _raise_dns_failure(url: str, host: str, exc: socket.gaierror) -> None:
    raise SSRFError(url, f"dns_failed: {exc}", host=host) from exc


def resolve_host_or_raise(url: str, host: str) -> str:
    """Synchronously resolve ``host`` and reject private answer records.

    Returns the first resolved address. The ``url`` is only used for
    error reporting / log context.

    DNS-rebinding posture: we inspect *every* address in each successful
    answer. An attacker who flips the authoritative
    DNS between this call and the eventual socket connect can still
    swing the answer — Python's stdlib has no zero-cost way to pin
    httpx's underlying ``httpcore.AsyncHTTPConnection`` to a captured
    IP. The first-line defence is the host-resolution check (catches
    almost every accidental misconfiguration plus a static-DNS
    attack); the second-line defence is the network firewall on the
    Hetzner box (port 5432 / 6379 / 8108 accept only the private
    network, not the public Internet). The Murmur shim's TS sibling
    pins the captured IP because Node exposes the ``lookup`` hook;
    httpx does not. If httpx ever exposes one, lift it from
    ``apps/murmur-shim/src/lib/murmur/ssrf.ts``.
    """
    try:
        infos = socket.getaddrinfo(host, None)
    except socket.gaierror as exc:
        _raise_dns_failure(url, host, exc)
    return _public_address_from_infos(url, host, infos)


async def _resolve_host_or_raise_with_retry(url: str, host: str) -> str:
    """Resolve off-loop and retry only temporary resolver unavailability."""
    max_attempts = len(DNS_RETRY_BACKOFF_SECONDS) + 1
    for attempt in range(1, max_attempts + 1):
        try:
            infos = await asyncio.to_thread(socket.getaddrinfo, host, None)
        except socket.gaierror as exc:
            if exc.errno != socket.EAI_AGAIN or attempt == max_attempts:
                _raise_dns_failure(url, host, exc)
            log.warning(
                "ssrf.dns.retry",
                host=host,
                attempt=attempt,
                max_attempts=max_attempts,
                error_type=type(exc).__name__,
            )
            await asyncio.sleep(DNS_RETRY_BACKOFF_SECONDS[attempt - 1])
            continue
        return _public_address_from_infos(url, host, infos)
    raise AssertionError("bounded DNS retry loop exhausted without a result")


def _host_requiring_dns_validation(url: str | httpx.URL) -> tuple[str, str] | None:
    """Validate URL shape/literals and return a hostname that needs DNS."""
    url_str = str(url)
    try:
        parsed = urlparse(url_str)
    except ValueError as exc:
        raise SSRFError(url_str, f"parse_failed: {exc}") from exc

    scheme = (parsed.scheme or "").lower()
    if scheme not in ("http", "https"):
        raise SSRFError(url_str, f"scheme_not_allowed:{scheme or 'empty'}")

    host = (parsed.hostname or "").lower()
    if not host:
        raise SSRFError(url_str, "host_missing")

    if is_internal_host_allowed(host):
        return None

    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        return url_str, host

    reason = _ip_in_blocklist(ip)
    if reason is not None:
        raise SSRFError(url_str, f"private_ip_literal:{reason}", host=host)
    return None


def validate_request_url(url: str | httpx.URL) -> None:
    """Public guard. Raises :class:`SSRFError` when ``url`` would SSRF.

    Called from :class:`SSRFGuardedTransport` for every request that
    httpx dispatches (including each redirect hop). Also exported for
    callers that want to fail-fast before queuing work.
    """
    target = _host_requiring_dns_validation(url)
    if target is not None:
        resolve_host_or_raise(*target)


async def _validate_request_url_with_dns_retry(url: str | httpx.URL) -> None:
    target = _host_requiring_dns_validation(url)
    if target is not None:
        await _resolve_host_or_raise_with_retry(*target)


class SSRFGuardedTransport(httpx.AsyncBaseTransport):
    """httpx transport wrapper that runs the SSRF guard per request.

    httpx calls ``handle_async_request`` once per request and **once
    per redirect hop** (see ``AsyncClient._send_handling_redirects``).
    That makes the transport the right place to enforce the guard:
    a single hook catches the initial connection AND every redirect
    target, with no need to thread the check through dozens of
    monitor / scraper call sites.

    The wrapper delegates to an inner transport (typically
    :class:`httpx.AsyncHTTPTransport`) for the actual fetch.
    """

    def __init__(self, inner: httpx.AsyncBaseTransport) -> None:
        self._inner = inner

    async def handle_async_request(self, request: httpx.Request) -> httpx.Response:
        try:
            await _validate_request_url_with_dns_retry(request.url)
        except SSRFError as exc:
            log.warning(
                "ssrf.guard.blocked",
                url=str(request.url),
                host=exc.host,
                reason=exc.reason,
            )
            raise
        return await self._inner.handle_async_request(request)

    async def aclose(self) -> None:
        await self._inner.aclose()


def wrap_transport_for_kwargs(
    kwargs: dict[str, Any],
    *,
    verify: Any = True,
) -> dict[str, Any]:
    """Patch a ``httpx.AsyncClient(**kwargs)`` dict to install the guard.

    The wrapper preserves any caller-supplied ``transport``; otherwise
    it constructs the standard :class:`httpx.AsyncHTTPTransport` with
    the same ``verify`` setting and wraps it.

    Returns the same dict for chaining.
    """
    inner = kwargs.pop("transport", None)
    if inner is None:
        inner = httpx.AsyncHTTPTransport(verify=verify)
    kwargs["transport"] = SSRFGuardedTransport(inner)
    return kwargs


def collect_internal_hosts() -> Iterable[str]:
    """Debug helper — return the current internal-host allowlist."""
    return sorted(_gather_internal_hosts())
